crates/iceberg/src/spec/table_metadata.rs

```diff
@@ -93,22 +112,291 @@ pub struct TableMetadata {
+            .iter()
+            .map(|field| field.field_id)
+            .max();
+        if max_id > self.last_partition_id {
+            self.last_partition_id = max_id;
+        }
+        if let Some(map) = self.partition_specs.as_mut() {
+            map.insert(partition_spec.spec_id, partition_spec);
+        } else {
+            self.partition_specs = Some(HashMap::from([(partition_spec.spec_id, partition_spec)]));
+        };
+        self
+    }
+
+    /// Add a snapshot to the TableMetadata and update last_sequence_number
+    /// snapshot : Snapshot to be added or replaced
+    /// current : True if the snapshot is the current one
+    pub fn with_snapshot(&mut self, snapshot: Snapshot, current: bool) -> &mut Self {
+        if current {
+            self.current_snapshot_id = Some(Some(snapshot.snapshot_id()))
+        }
+        if Some(snapshot.sequence_number()) > self.last_sequence_number {
+            self.last_sequence_number = Some(snapshot.sequence_number())
+        }
+        if let Some(Some(map)) = self.snapshots.as_mut() {
+            map.insert(snapshot.snapshot_id(), Arc::new(snapshot));
+        } else {
+            self.snapshots = Some(Some(HashMap::from([(
+                snapshot.snapshot_id(),
+                Arc::new(snapshot),
+            )])));
+        };
+
+        self
+    }
+
+    /// Add a sort_order to the TableMetadata
+    /// sort_order : SortOrder to be added or replaced
+    /// default: True if this SortOrder is the default one
+    pub fn with_sort_order(&mut self, sort_order: SortOrder, default: bool) -> &mut Self {
+        if default {
+            self.default_sort_order_id = Some(sort_order.order_id)
+        }
+        if let Some(map) = self.sort_orders.as_mut() {
+            map.insert(sort_order.order_id, sort_order);
+        } else {
+            self.sort_orders = Some(HashMap::from([(sort_order.order_id, sort_order)]));
+        };
+        self
+    }
+
+    /// Add a snapshot_reference to the TableMetadata
+    /// key_ref : reference id of the snapshot
+    /// snapshot_ref : SnapshotReference to add or update
+    pub fn with_ref(&mut self, key_ref: String, snapshot_ref: SnapshotReference) -> &mut Self {
+        if let Some(map) = self.refs.as_mut() {
+            map.insert(key_ref, snapshot_ref);
+        } else {
+            self.refs = Some(HashMap::from([(key_ref, snapshot_ref)]));
+        };
+        self
+    }
+
+    /// Check if the default key exists in the map.
+    /// Verify incoherent behavior and throw Error if :
+    /// - the map is not defined but the key exists
+    /// - the default key exists but there is no map
+    ///     except if key is a default value
+    /// Params :
+    /// - key : the key to look for in the map
+    /// - map : the map to scan
+    /// - default : default value for the key if any
+    /// - field : map name for throwing a better error
+    fn check_id_in_map<K: std::hash::Hash + Eq + std::fmt::Display + Copy, T>(
```

Comment:
Do we really need to check this at last? Why not check it during modification?

Reply:
if we restrict with only the  given list of setter below we should not need this kind of validation as we will compute every value inside the builder. So they should be coherent.